runtime/gc_glue_java/ConfigurationDelegate.hpp

```diff
@@ -93,13 +91,14 @@ class MM_ConfigurationDelegate
 		javaVM->gcWriteBarrierType = writeBarrierType;
 
 		if (extensions->alwaysCallReadBarrier) {
+			/* AlwaysCallReadBarrier takes precedence over other read barrier types */
 			javaVM->gcReadBarrierType = gc_modron_readbar_always;
+		} else if (extensions->isConcurrentScavengerEnabled()) {
+			javaVM->gcReadBarrierType = gc_modron_readbar_range_check;
+		} else if (extensions->isConcurrentCopyForwardEnabled()) {
+			javaVM->gcReadBarrierType = gc_modron_readbar_region_check;
 		} else {
-			if (extensions->isConcurrentScavengerEnabled()) {
-				javaVM->gcReadBarrierType = gc_modron_readbar_range_check;
-			} else {
-				javaVM->gcReadBarrierType = gc_modron_readbar_none;
-			}
+			javaVM->gcReadBarrierType = gc_modron_readbar_none;
```

Comment:
Now that you are touching this could, how about we restrict these with GC policy (which is easily accessibly here), as well?

Imagine somebody runs Balanced and provides  options for both Concurrent Copy Forward and (mistakenly) Concurrent Scavenger. With the current logic, we would set the barrier type to range_check, what is incorrect.

Reply:
Yeah, it would make more sense to set it based on the GC Policy.  I'm thinking about setting the read barrier from the `MM_Configuration` implementation used, which is how the write barrier is set. Does that seem reasonable?

Reply:
I went the easy root for now, and just checked the current GC policy in place.  Let me know what you think is best.

Reply:
We could based it on _gcPolicy, but this  is equally good to me.